### PR TITLE
enable qlora finetuning on single GPU

### PR DESCRIPTION
**Pull request by weifengpy**
```
python transformer_nuggets/llama/finetune.py
```

Adds qlora training option. Next step will be enable torchrun + FSDP

### CODE REVIEW

File: transformer_nuggets/llama/train.py [hunk 1]
```diff
@@ -15,7 +15,8 @@
 import numpy as np
 import torch
 from fire import Fire
-from float8_experimental.dynamic_linear import Float8DynamicLinear
```
Comment (weifengpy) [inline transformer_nuggets/llama/train.py:18]: seems there is a name change in trunk: https://github.com/pytorch-labs/float8_experimental/blob/main/float8_experimental/float8_dynamic_linear.py

Comment (drisspg) [reply transformer_nuggets/llama/train.py:18]: totally was, thanks for fixing !

File: transformer_nuggets/llama/train.py [hunk 2]
```diff
@@ -82,6 +83,11 @@ class TrainingConfig:
     # on a single batch. It should be set to False for normal training.
     overfit: bool = False
 
+    enable_qlora: bool = False
+    lora_r: int = 8
```
Comment (drisspg) [inline transformer_nuggets/llama/train.py:87]: Nit: TrainingConfig I think these should maybe go in HyperParams, tried to have logical distitincation between, "effects individual training run" and then hyperparams "effect the model arch" and may want to be searched over

but either way is cool those have already gotten a little blurry

File: transformer_nuggets/quant/qlora.py [hunk 3]
```diff
@@ -439,3 +441,44 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        self.qlora_w2 = QloraLinear(weight2.shape[1], weight2.shape[0], weight2, lora_r, lora_alpha, lora_dropout)
+        self.qlora_w3 = QloraLinear(weight3.shape[1], weight3.shape[0], weight3, lora_r, lora_alpha, lora_dropout)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = F.silu(self.qlora_w1(x)) * self.qlora_w2(x)
+        x = self.qlora_w3(x)
+        return x
+
+def swap_for_qlora_jank(model: torch.nn.Module, qlora_config: QloraConfig, dtype) -> None:
+    print("Swapping for Qlora...")
```
Comment (drisspg) [inline transformer_nuggets/quant/qlora.py:473]: can you logging.info for this

**Review by drisspg — APPROVED** [review]
Looks good the one overall nit is maybe we just copy and paste this into a seperate "finetune.py" file since I think that will make it easier going forward since there is subtlety on how you init the model

File: transformer_nuggets/quant/qlora.py [hunk 4]
```diff
@@ -439,3 +441,44 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        self.qlora_w1 = QloraLinear(weight1.shape[1], weight1.shape[0], weight1, lora_r, lora_alpha, lora_dropout)
+        self.qlora_w2 = QloraLinear(weight2.shape[1], weight2.shape[0], weight2, lora_r, lora_alpha, lora_dropout)
+        self.qlora_w3 = QloraLinear(weight3.shape[1], weight3.shape[0], weight3, lora_r, lora_alpha, lora_dropout)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = F.silu(self.qlora_w1(x)) * self.qlora_w2(x)
+        x = self.qlora_w3(x)
+        return x
+
+def swap_for_qlora_jank(model: torch.nn.Module, qlora_config: QloraConfig, dtype) -> None:
```
Comment (awgu) [inline transformer_nuggets/quant/qlora.py:472]: What would the ideal state look like (if this way is the 'jank' way)?

File: transformer_nuggets/quant/qlora.py [hunk 5]
```diff
@@ -439,3 +441,44 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        feed_forward = module.feed_forward
+        # it's not a typo to match QloraMLP.w2 with FeedForward.w3
+        # FeedForward forward: w2(silu(w1) * w3)
+        # QloraMLP forward: w3(silu(w1) * w2)
+        w1 = feed_forward.w1.weight.to(dtype=dtype)
+        w2 = feed_forward.w3.weight.to(dtype=dtype)
+        w3 = feed_forward.w2.weight.to(dtype=dtype)
+        new_mod = QloraMLP(w1, w2, w3, qlora_config)
+        module.feed_forward = new_mod
+        del feed_forward
```
Comment (awgu) [inline transformer_nuggets/quant/qlora.py:484]: nit: I do not think this `del` is needed. `feed_forward` gets overridden immediately in the next loop iteration for all but the last iteration, and in the last iteration, it gets GC'd upon the function return.

File: transformer_nuggets/llama/train.py [hunk 6]
```diff
@@ -158,10 +164,22 @@ def main(
...
+        if training_config.enable_qlora:
+            qlora_config = qlora.QloraConfig(
+                training_config.lora_r,
+                training_config.lora_alpha,
+                training_config.lora_dropout,
+            )
+            qlora.swap_for_qlora_jank(model, qlora_config, torch.bfloat16)
+            for name, param in model.named_parameters():
+                if "lora_" not in name:
+                    param.requires_grad = False
```
Comment (awgu) [inline transformer_nuggets/llama/train.py:181]: nit: Should this freezing happen in the `swap_for_qlora` function?

File: transformer_nuggets/quant/qlora.py (hunk 5 above)
Comment (drisspg) [reply transformer_nuggets/quant/qlora.py:484]: yeah this was me being overly cautious

File: transformer_nuggets/quant/qlora.py [hunk 7]
```diff
@@ -439,3 +441,44 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = F.silu(self.qlora_w1(x)) * self.qlora_w2(x)
+        x = self.qlora_w3(x)
+        return x
+
+def swap_for_qlora_jank(model: torch.nn.Module, qlora_config: QloraConfig, dtype) -> None:
+    print("Swapping for Qlora...")
+    for module in tqdm(model.layers):
+        feed_forward = module.feed_forward
+        # it's not a typo to match QloraMLP.w2 with FeedForward.w3
```
Comment (awgu) [inline transformer_nuggets/quant/qlora.py:476]: Curious why not change `QloraMLP` to match `FeedForward`?

Comment (weifengpy) [reply transformer_nuggets/quant/qlora.py:476]: good point, I can change ``QloraMLP``. I

File: transformer_nuggets/quant/qlora.py (hunk 4 above)
Comment (weifengpy) [reply transformer_nuggets/quant/qlora.py:472]: will remove "_jank" from the name

File: transformer_nuggets/llama/train.py [hunk 8]
```diff
@@ -15,7 +15,8 @@
 import numpy as np
 import torch
 from fire import Fire
-from float8_experimental.dynamic_linear import Float8DynamicLinear
+import transformer_nuggets.quant.qlora as qlora
```
Comment (drisspg) [inline transformer_nuggets/llama/train.py:18]: can we remove, this and other mentions of qlora in the train.py

**Comment by drisspg** [comment]
also one other thing, is if yo install the [dev] dependencies it shoudl install pre-commit hooks that will autoformat for you but can you run "ufmt format ." on all of this

**Pull request merged**